### DISCUSSION TITLE
Feature types and explainability

### DISCUSSION THREAD

**Discussion by Yasslight90** [microsoft/FLAML#220] [Q&A]
Hi there,

I have three questions: 

1- What feature types does FLAML support? I am sure it supports at least numerical and categorical features. Any other types?
2- Is it possible to force the type for a given feature? 
3- Does FLAML provide an explanation for predictions? I am thinking of shap values for example. 

Thank you so much

Yassine - Data scientist

**Comment by sonichi** [comment]
1. numerical, categorical and datetime.
2. if you pass a pandas dataframe, the type is specified for each column. 
3. explanation is not built-in from flaml. Is there any package you are using for explanation now?

**Comment by Yasslight90** [comment]
Thank you for the quick answer!

3- No. I am evaluating the autoML-ness of many solutions and regarding their explainability I found out that they all use SHAP values to explain predictions.

**Reply by sonichi** [reply]
Thanks for the discussion. I'd like to get your opinion on: should explanation of predictions be part of an automl library? If so, what is the right thing to be automated?

**Comment by Yasslight90** [comment]
Hello,

Yes. From my point of view, autoML solutions should definetely be able to explain their outputs, especially when used in critical applications:  Today, most of the existing solutions (Google autoML tables, Prevision.io, H2O, etc) use shapley values which are not very trustworthy. 

Moreover, what I think should be provided to the user is: 1) a global interpretation (for example, by providing him/her with a feature importance plot) and 2) a local interpretation (by computing the importance of each feature in every prediction).

Cheers

Yassine

**Reply by sonichi** [reply]
Thanks for the input. It'll be nice to have a real use case where this feature is requested. If you have one, I'll be happy to hear it. Feel free to chat on [gitter](https://gitter.im/FLAMLer/community#) if you like.

**Reply by Yasslight90** [reply]
a simple use case would be: a user asking for a bank loan. Imagine that the bank relies on the predictions of a automated machine learning model to accept or not a bank loan. Clients who got a refusal would most likely want to understand this decision and the bank owe them an explanation; here we are talking about a local explanation. Also, bank advisor would want to know what are the most important variables in the classification (acceptance, refusal); here we are talking about a global interpretation.

**Reply by kevball2** [reply]
Are you looking for something like this? https://github.com/microsoft/responsible-ai-toolbox

Azure AutoML has some of these features in preview 
https://docs.microsoft.com/en-us/azure/machine-learning/how-to-machine-learning-fairness-aml

**Reply by Yasslight90** [reply]
Thanks. However, I am not looking for a solution for explainability. I am just highliting the importance of having it integrated into the autoML process :-). For example, you would integrate Fairlearn into FLAML ...

**Comment by qingyun-wu** [comment]
Hi @Yasslight90 and @kevball2,

Thank you for your insightful comments. There are several related TODOs included in our roadmap:
- [Integrate Fair AutoML, which leverages FairLearn to improve the fairness of the resulting ML models from FLAML](https://github.com/microsoft/FLAML/wiki/Roadmap-for-Upcoming-Features#fair-automl). This is ongoing research. We plan to do the integration in the near future. Feel free to check out our paper and share your comments.
- [Prediction quality](https://github.com/microsoft/FLAML/wiki/Roadmap-for-Upcoming-Features#prediction-quality). I believe this also contributes to improving explanation. The idea is to report the uncertainty of a model's predction in addition to the accuracy, e.g., via conformal prediction.
- [Visualizing feature importance, explanation, and optimization history](https://github.com/microsoft/FLAML/wiki/Roadmap-for-Upcoming-Features#visualize-feature-importance-shaplime-explanation-optimization-history). This is still in an early stage investigating what should be included. Here are my understandings:
             - [Lime](https://github.com/marcotcr/lime) can provide local explanations (e.g., feature importance for specific instances) as shown in this [example](https://github.com/marcotcr/lime/blob/master/doc/notebooks/Lime%20-%20basic%20usage%2C%20two%20class%20case.ipynb).
             - [Shap](https://github.com/slundberg/shap) can provide both local and global explanations via shap values.
             And for integrating them into the AutoML process: a naive integration I can think of is just to apply these explainers (and the 
             confidence level) **at the end of the AutoML process** to the resulting model (the best model) of FLAML. Do you have additional comments on where this explanation should happen in the AutoML process?  


Thank you!
